pkg/certificate/providers/mrcclient.go

```diff
@@ -0,0 +1,77 @@
+package providers
+
+import (
+	"context"
+
+	"github.com/openservicemesh/osm/pkg/apis/config/v1alpha2"
+	"github.com/openservicemesh/osm/pkg/certificate"
+	"github.com/openservicemesh/osm/pkg/k8s/informers"
+	"github.com/pkg/errors"
+	"k8s.io/client-go/tools/cache"
+)
+
+var (
+	errNilInformerCollection = errors.New("informer collection is nil")
+)
+
+// TODO: Rename (keithmattix)
+type MRCClientImpl struct {
+	informerCollection *informers.InformerCollection
+	MRCProviderGenerator
+}
+
+func (m *MRCClientImpl) List() ([]*v1alpha2.MeshRootCertificate, error) {
+	if m.informerCollection == nil {
+		return nil, errNilInformerCollection
+	}
+	// informers return slice of pointers so we'll convert them to value types before returning
+	mrcPtrs := m.informerCollection.List(informers.InformerKeyMeshRootCertificate)
+	var mrcs []*v1alpha2.MeshRootCertificate
+	for _, mrcPtr := range mrcPtrs {
+		if mrcPtr == nil {
+			continue
+		}
+		mrc, ok := mrcPtr.(*v1alpha2.MeshRootCertificate)
+		if !ok {
+			continue
+		}
+		mrcs = append(mrcs, mrc)
+	}
+
+	return mrcs, nil
+}
+
+func (m *MRCClientImpl) Watch(ctx context.Context) (<-chan certificate.MRCEvent, error) {
+	if m.informerCollection == nil {
+		return nil, errNilInformerCollection
+	}
+
+	eventChan := make(chan certificate.MRCEvent)
+	m.informerCollection.AddEventHandler(informers.InformerKeyMeshRootCertificate, cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			mrc := obj.(*v1alpha2.MeshRootCertificate)
+			eventChan <- certificate.MRCEvent{
+				Type: certificate.MRCEventAdded,
+				MRC:  mrc,
+			}
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			oldMRC := oldObj.(*v1alpha2.MeshRootCertificate)
+			mrc := newObj.(*v1alpha2.MeshRootCertificate)
+			eventChan <- certificate.MRCEvent{
+				Type:   certificate.MRCEventUpdated,
+				MRC:    mrc,
+				OldMRC: oldMRC,
+			}
+		},
+		DeleteFunc: func(obj interface{}) {
```

Comment:
Not sure we need this. But I don't think it hurts

Reply:
You have to provide a Delete function otherwise, you'll get a nil pointer exception when the internal informer logic tries to handle the event.